modules/Bio/EnsEMBL/Compara/Utils/TaxonomicReferenceSelector.pm

```diff
@@ -0,0 +1,143 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+
+=head1 NAME
+
+Bio::EnsEMBL::Compara::Utils::TaxonomicReferenceSelector
+
+=head1 DESCRIPTION
+
+Utility methods for collecting taxonomic ranks and matching to taxonomic references
+
+=cut
+
+package Bio::EnsEMBL::Compara::Utils::TaxonomicReferenceSelector;
+
+use strict;
+use warnings;
+use base qw(Exporter);
+use Bio::EnsEMBL::Utils::Exception qw(throw);
+use Bio::EnsEMBL::Compara::GenomeDB;
+use Bio::EnsEMBL::Compara::DBSQL::DBAdaptor;;
+use List::Util;
+
+our %EXPORT_TAGS;
+our @EXPORT_OK;
+
+@EXPORT_OK = qw(
+    collect_reference_classification
+    match_query_to_reference_taxonomy
+    collect_species_set_dirs
+);
+%EXPORT_TAGS = (
+  all     => [@EXPORT_OK]
+);
+
+=head2 collect_reference_classification
+
+    Arg[1]     :  (string) $master_db or Bio::EnsEMBL::Compara::DBSQL::DBAdaptor $master_db
+    Description:  Collect reference ncbi taxonomic classifications by species_set. This expects a reference
+                  specific compara master database with taxonomically named species_sets.
+                  E.g. \@taxon_clade = collect_reference_classification('master_db');
+    Return     :  (arrayref) list of species_set names.
+    Exceptions :  None.
+
+=cut
+
+sub collect_reference_classification {
+    my $master_db = shift;
+
+    my $dba = Bio::EnsEMBL::Compara::DBSQL::DBAdaptor->go_figure_compara_dba($master_db);
+    my $ss_adaptor = $dba->get_SpeciesSetAdaptor->fetch_all();
+    my @ss_names   = map {$_->name()} @{$ss_adaptor};
+    my @clean_ss   = grep ( s/^collection\-//g, @ss_names );
+    @clean_ss      = sort @clean_ss;
+    return \@clean_ss;
+}
+
+=head2 match_query_to_reference_taxonomy
+
+    Arg[1]     :  Bio::EnsEMBL::Compara::DBSQL::DBAdaptor $compara_dba
+    Arg[2]     :  Bio::EnsEMBL::Compara::GenomeDB
+    Arg[3]     :  (optional) Bio::EnsEMBL::Compara::DBSQL::DBAdaptor $master_dba
+                  Either/or
+    Arg[4]     :  (optional) (arrayref) $taxon_list
+    Description:  Match starting from lowest taxonomic rank climbing, returning string variable at first match.
+                  E.g. my $ref_clade = match_query_to_reference_taxonomy($compara_dba, $genome_db);
+    Return     :  (string) $taxon_name or undef
+    Exceptions :  Throws if both $taxon_list and $master_dba provided
+
+=cut
+
+sub match_query_to_reference_taxonomy {
+    my ($genome_db, $master_dba, $taxon_list) = (@_);
+
+    throw ("taxon_list and master_dba are mutually exclusive, pick one") if $taxon_list && $master_dba;
+    throw ("Either taxon_list or master_dba need to be provided") unless $taxon_list || $master_dba;
+
+    my @taxon_list = $taxon_list ? @$taxon_list : @{collect_reference_classification($master_dba)};
+    my $parent     = $genome_db->taxon->parent;
+
+    while ( $parent->name ne "root" ) {
+        if ( grep { lc($parent->name) eq $_ } @taxon_list ) {
```

Comment:
Not sure how many reference collections we're anticipating in the future, but making `@taxon_list` into a hash would be a faster way to check for existence than `grep`